test/test_datapipe.py

```diff
@@ -2290,5 +2296,162 @@ def test_old_dataloader(self):
         self.assertEqual(sorted(expected), sorted(items))
 
 
+class TestIterDataPipeSingletonConstraint(TestCase):
+
+    r"""
+    Each `IterDataPipe` can only have one active iterator. Whenever a new iterator is created, older
+    iterators are invalidated. These tests aim to ensure `IterDataPipe` follows this behavior.
+    """
+
+    def test_iterdatapipe_singleton_generator(self):
+        r"""
+        Testing for the case where IterDataPipe's `__iter__` is a generator function.
+        """
+        # Functional Test: There should be an error with multiple iterators
+        source_dp: IterDataPipe = dp.iter.IterableWrapper(range(10))
+        it1 = iter(source_dp)
+        self.assertEqual(list(range(10)), list(it1))  # It can
+        it1 = iter(source_dp)
+        self.assertEqual(list(range(10)), list(it1))
+        it1 = iter(source_dp)
+        self.assertEqual(0, next(it1))
+        it2 = iter(source_dp)
+        self.assertEqual(0, next(it2))
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            next(it1)
+
+        # Functional Test: extend the test to a pipeline
+        dps = source_dp.map(_fake_fn).filter(_fake_filter_fn)
+        it = iter(dps)
+        self.assertEqual(list(range(10)), list(it))
+
+        # Functional Test: multiple simultaneous references to the same DataPipe fails
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            for _ in zip(source_dp, source_dp):
+                pass
+
+        # Function Test: sequential references work
+        for _ in zip(list(source_dp), list(source_dp)):
+            pass
+
+    def test_iterdatapipe_singleton_self_next(self):
+        r"""
+        Testing for the case where IterDataPipe's `__iter__` returns `self` and there is a `__next__` method
+        Note that the following DataPipe by is singleton by default (because `__iter__` returns `self`).
+        """
+        class _CustomIterDP_Self(IterDataPipe):
+            def __init__(self, iterable):
+                self.source = iterable
+                self.iterable = iter(iterable)
+
+            def __iter__(self):
+                self.reset()
+                return self
+
+            def __next__(self):
+                return next(self.iterable)
+
+            def reset(self):
+                self.iterable = iter(self.source)
+
+        # Functional Test: Check that every `__iter__` call returns the same object
+        source_dp = _CustomIterDP_Self(range(10))
```

Comment:
Amazing work! The only thing missing by my opinion is moving all tests commonalities into something like:
`def _test_datapipe_iterators_restrictions(self, dp_graph)`

Reply:
I have refactored the test.